# Python Basics: Assignment 1 - Solutions

## Section 1: Numerical Data Types & Arithmetic Operators

In [1]:
fuel_efficiency = 11.5  # km per liter
trip_distance = 475     # kilometers
liters_needed = trip_distance / fuel_efficiency
print("Liters needed:", liters_needed)

Liters needed: 41.30434782608695


**Output:** 41.30

**Explanation:** Total distance divided by fuel efficiency.

In [ ]:
tank_capacity = 50
full_tanks = int(liters_needed // tank_capacity)
print("Full tanks:", full_tanks)

**Output:** 0

**Explanation:** Floor division gives whole number of full tanks.

In [ ]:
fuel_price = 1.65
fuel_cost = liters_needed * fuel_price
print("Fuel cost: $", fuel_cost)

**Output:** 68.15

**Explanation:** Multiply liters needed by price per liter.

In [ ]:
fuel_remaining = tank_capacity - liters_needed
print("Fuel remaining:", fuel_remaining)

**Output:** 8.69

**Explanation:** Subtract fuel used from tank capacity.

## Section 2: Comparison & Logical Operators

In [ ]:
age = 30
is_weekday = True
is_family_day = False
group_size = 3
discount = (age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
print("Discount (initial):", discount)

**Output:** False

In [ ]:
age = 10
is_weekday = True
is_family_day = False
group_size = 2
discount = (age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
print("Discount (new):", discount)

**Output:** True

**Step-by-step:**
- (age <= 12 and is_weekday) → True
- (age >= 65) → False
- (is_family_day and group_size >= 4) → False
- Final Result: True

## Section 3: String Data Types, Indexing & Slicing

In [ ]:
secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"
print(secret_message[5])

**Output:** N

In [ ]:
print(secret_message[7:13])

**Output:** Is_FuN

In [ ]:
print(secret_message[-10:-1])

**Output:** PoWeRfUl

In [ ]:
print(secret_message[0::4])

**Output:** PoNFNDU

```python
# print(secret_message[50])
```
Would raise: IndexError

**Explanation:** Index exceeds string length.

## Section 4: Boolean Values & Expressions

In [ ]:
x = 15
y = 25
z = 15
text = "adventure"
print((x == z) and (y > x))

**Output:** True

In [ ]:
print((x != z) or (y < x) or (len(text) < 5))

**Output:** False

In [ ]:
print(not ((x > y) and (text == "Adventure")))

**Output:** True

## Section 5: String Properties & Basic Methods

```python
username = "MyUser123"
username[0] = 'm'  # Error
```
**Output:** TypeError

**Explanation:** Strings are immutable.

In [ ]:
quote = " practice makes perfect "
formatted_quote = quote.strip().upper().replace("PERFECT", "PROGRESS")
print("Formatted quote:", formatted_quote)

**Output:** PRACTICE MAKES PROGRESS

In [ ]:
print("Original length:", len(quote))
print("Formatted length:", len(formatted_quote))

**Output:** 25, 24